optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/AbstractSolver.java

```diff
@@ -100,20 +109,62 @@ public void solvingEnded(DefaultSolverScope<Solution_> solverScope) {
         bestSolutionRecaller.solvingEnded(solverScope);
         termination.solvingEnded(solverScope);
         phaseLifecycleSupport.fireSolvingEnded(solverScope);
+        solverScope.endingNow();
+        solverListenerSupport.fireSolvingEnded(solverScope);
     }
 
     // ************************************************************************
     // Event listeners
     // ************************************************************************
 
     @Override
-    public void addEventListener(SolverEventListener<Solution_> eventListener) {
-        solverEventSupport.addEventListener(eventListener);
+    public void addSolverListener(SolverListener<Solution_> listener) {
+        if (isSolving()) {
```

Comment:
Thinking aloud: this might be a race condition when the `isSolving()` method is false in a thread adding a listener and true in a thread running the solver. But everyone should register a listener before starting a solver, so it should be ok.

Reply:
Yes, there is a race condition that this fail-fast might not be detected in very race cases. The method says it's NOT trade safe, so anyone doing this is effectively violating the javadoc contract (which is exactly what we're detecting most of the time with this fail fast).